793 route path link modification

Divinare opened the pull request:
Closes #793 

BE PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/58

Node saving will be done at task: #859

- Now textContainer displays "-" as empty value.
- InputContainer can take null, just so that you dont have to say: x ? x : undefined, you can give null straigtly
- routePath without routePathLinks can now be opened
- extendRoutePath tool now puts on node & link layers
- fixed undo / redo to update only routePath geometries

kronken commented on src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx:
```diff
@@ -21,11 +23,87 @@ interface IRoutePathListLinkProps {
+        if (
+            prevProps.isEditingDisabled !== this.props.isEditingDisabled &&
+            !this.props.isEditingDisabled
+        ) {
```
if (prevProps.isEditingDisabled && !this.props.isEditingDisabled) {}

kronken commented on src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx:
```diff
@@ -21,11 +23,87 @@ interface IRoutePathListLinkProps {
+    private onRoutePathLinkPropertyChange = (property: string) => (
+        value: any
+    ) => {
+        const _value = value === '' ? null : value;
```
I assume this is done in order to prevent empty strings in the database. If it is indeed for that reason, then I don't think this is the place where that logic should happen.

Divinare replied:
I guess this was because I thought startNodeColumnNumber didn't quite work when it was ''. Not it works, maybe because of other bug fixes --> i removed this line

kronken commented on src/models/IRoutePathLink.ts:
```diff
@@ -8,16 +8,21 @@ interface IRoutePathLinkPrimaryKey {
     startNodeType: NodeType;
     isStartNodeTimeAlignmentStop: boolean;
+    isStartNodeHastusStop: boolean;
+    isAtBookSchedule: boolean;
```
isUsingSchedulesFromBook
?

kronken replied:
or
isStartNodeUsingScheludesFromBook

Divinare replied:
isStartNodeAtBookSchedule

isPassTimeOfStartNodeAtBookSchedule

Divinare replied:
isStartNodeInBookSchedule

Divinare replied:
isStartNodePassTimeInBookSchedule

Divinare replied:
isStartNodeUsingBookSchedule

kronken commented on src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx:
```diff
@@ -86,91 +167,49 @@ class RoutePathListLink extends React.Component<IRoutePathListLinkProps> {
-                    </div>
+                    <InputContainer
+                        disabled={isEditingDisabled}
+                        type={'number'}
```
No need to wrap string with {}

Divinare replied:
Also found some other places to fix this with ctrl + F

kronken commented on src/models/IRoutePathLink.ts:
```diff
@@ -8,16 +8,21 @@ interface IRoutePathLinkPrimaryKey {
     isStartNodeTimeAlignmentStop: boolean;
+    isStartNodeHastusStop: boolean;
+    isAtBookSchedule: boolean;
+    startNodeColumnNumber: number | null;
```
startNodeBookScheduleColumnNumber
?

Divinare replied:
startNodeBookColumnNumber

kronken commented on src/models/validationModels/routePathLinkValidationModel.ts:
```diff
@@ -0,0 +1,16 @@
+    isStartNodeTimeAlignmentStop: 'boolean',
+    isStartNodeHastusStop: 'boolean',
+    isAtBookSchedule: 'boolean',
+    startNodeColumnNumber: 'numeric|max:99'
```
min 0 ?

kronken replied:
Saving a large negative value is currently broken, and I don't see any reason why we should allow negative numbers

Divinare commented:
Review fixes done.

In addition, I made some changes:

- remove font-weight: bold from .list, inputLabels are now also darker
   * now the look of links & nodes tab is a lot nicer if compared to old version
- removed "Solmun tiedot", "Pysäkin tiedot", "Solmun tyyppi" (because 
- For "sarakenumero should be increasing number" I created a new card: #864
- added checkbox (that does nothing at the moment):
[]  Pysäkki ei käytössä, relpysäkki = 'P'
[X] Pysäkki ei käytössä, relpysäkki = 'E'
- disable kirjasarake input when checkbox is checked

kronken commented:
If two stops are after each other in routepath, it seems that updating the ohitusaika data for one, does also update the other

Divinare commented:
RoutePathList item styling TODO at task #870

kronken commented:
There is still some problem when the routepath has two stops in the end. the logic seems to believe that both stops are the last stop, and modifying one does also modify the other

Divinare commented:
Fixed bug with 2 consecutive stops & bug undo/redo.

Can be reviewed again.
